report anonymous usage data (telemetry)

EverlastingBugstopper opened the pull request:
fixes #44 

an example request when running the command `rover config profile list`

```
body: 
  command:     config profile list
  machine_id:  edd890f0-3f8d-43f5-a22e-d3731d7e5042
  session_id:  bbc30ec1-628e-4563-83af-cb2bfe06e997
  cwd_hash:    52b120fee6f41776b6fb561ecb708cfe187e6922fa33a2399f8e01cb94e89bb0
  platform: 
    os:                     macos
    continuous_integration: null
  cli_version: 0.0.0
```

good thing about this approach is that any new commands should be automatically serialized and reported without any additional code or remembering to add the command to the telemetry reporter. the downside to this is that if there are arguments/fields that contain sensitive information that we don't want to collect, you must mark that field as `#[serde(skip_serializing)]` so that it is not included in the report.

all of this capturing happens on a background thread which should mitigate most performance issues that could occur if the reporting was run synchronously.

outstanding work

- [X] decide on capturing additional info (federated v non-federated)
  - **punting until later, not gonna do anything with this one right now**
- [X] figure out back end
  - **wip pr [here](https://github.com/apollographql/spaceport/pull/4)**
- [ ] make an opt-in serialization API to prevent accidentally capturing user-specified arguments
  - **waiting on a response [here](https://github.com/serde-rs/serde/issues/1897)**
- [x] document usage of `APOLLO_TELEMETRY_DISABLED` and the data we collect
- [x] add section to CONTRIBUTING.md for new folks :)
- [x] figure out how to display `command` nicely in Segment
    - ~should we just have the top level subcommand (`Config` in this example) be the `event` and then have the rest of the arguments be a part of `properties`?~ **yes**
    - ~should we do that part up front in Rover, or should we do it in the Worker?~ **this PR does it in Rover**

ashleygwilliams commented on Cargo.toml:
```diff
@@ -1,23 +1,25 @@
 [package]
-name = "rover"
```
this lint is not my fave (name should be at the top!!)

EverlastingBugstopper replied:
mmm y'know that is very fair. perhaps ill just get rid of it

ashleygwilliams commented on crates/sputnik/src/session.rs:
```diff
@@ -0,0 +1,136 @@
+use anyhow::Result;
+use ci_info::types::Vendor;
```
maybe `as CiVendor` to give vendor some context if we're bringing it to global scope

ashleygwilliams commented on src/cli.rs:
```diff
@@ -22,3 +29,39 @@ impl Rover {
     }
 }
+
+impl Report for Rover {
```
any chance i could persuade you to bump this out to its own file?

ashleygwilliams commented on src/cli.rs:
```diff
@@ -22,3 +29,39 @@ impl Rover {
+        if let Some(url) = option_env!("APOLLO_TELEMETRY_URL") {
+            Ok(url.parse()?)
+        } else if cfg!(debug_assertions) {
+            Ok("http://localhost:8000".parse()?)
```
would love to pull these out as consts

ashleygwilliams commented on src/cli.rs:
```diff
@@ -22,3 +29,39 @@ impl Rover {
+
+    fn version(&self) -> String {
+        option_env!("CARGO_PKG_VERSION")
+            .unwrap_or("0.0.0")
```
not sure i agree with this default?

EverlastingBugstopper replied:
yeah... wasnt sure about this one. think we should panic by default instead? or return an error?

ashleygwilliams reviewed (commented):
in general this LGTM, looking forward to the in person review, but a few nits beforehand

jbaxleyiii commented on crates/houston/src/error.rs:
```diff
@@ -0,0 +1,31 @@
+
+/// HoustonProblem is the type of Error that occured.
+#[derive(Error, Debug)]
+pub enum HoustonProblem {
```
👏🏼👏🏼👏🏼👏🏼👏🏼👏🏼👏🏼👏🏼👏🏼

jbaxleyiii commented on crates/houston/src/error.rs:
```diff
@@ -0,0 +1,31 @@
+#[derive(Error, Debug)]
+pub enum HoustonProblem {
+    /// ConfigDirNotFound occurs when the default OS config can't be found.
+    #[error("Could not determine default OS config directory.")]
```
Is there a way to make these messages more actionable?

EverlastingBugstopper replied:
the actionable error messages will have to come directly from the CLI implementation, I left them out of this crate so as to be less opinionated about the error messages displayed to the user, the CLI as the consumer should have control over those messages. This PR does not include those changes however.

jbaxleyiii commented on crates/houston/src/error.rs:
```diff
@@ -0,0 +1,31 @@
+
+    /// TomlSerialization occurs when a profile's configuration can't be serialized to a String.
+    #[error(transparent)]
+    TomlSerialization(#[from] toml::ser::Error),
```
I'm a little suprised to see these implemented as `HoustonErrors`. Is that just to get the `from` casting so that we can use `?`? I'm a little unsure of how well this will scale?

EverlastingBugstopper replied:
This is the way that `thiserror` recommends catching and returning errors and allows consumers of the library to know what types of error may arise within an application. This lets the crate keep a very clean API and does not require consumers of the library to import error types from the libraries that `Houston` uses under the hood. I'd recommend taking a look at [the docs for `thiserror`](https://docs.rs/thiserror/1.0.20/thiserror/)

jbaxleyiii commented on crates/sputnik/README.md:
````diff
@@ -0,0 +1,23 @@
+
+```sh
+body:
+  command:     {"command":{"Config":{"command":{"Profile":{"command":"List"}}}}}
````
This structure will be hard to search / index within our backend. Does this only capture arguments and not flags? If so, why not a `vec<String>` or something more simple where the depth isn't variable?

EverlastingBugstopper replied:
this will capture _anything_ that is part of our `structopt` fields that do not have `#[serde(skip_serializing)` on it, so it should capture _all_ information we want. we can also write custom seriazliers for different commands to override the defaults, though i do not have an example of how to do that which i can take as an action item

jbaxleyiii commented on crates/sputnik/src/error.rs:
```diff
@@ -0,0 +1,32 @@
+pub enum SputnikError {
+    /// IOError occurs when any given std::io::Error arises.
+    #[error(transparent)]
+    IOError(#[from] io::Error),
```
I have the same scaling questions about the `from` implementations here. Is this better than asking the calling code to do the conversion? I'm really not sure!

jbaxleyiii commented on crates/sputnik/src/report.rs:
```diff
@@ -0,0 +1,59 @@
+    /// machine identifier
+    fn machine_id_config(&self) -> Result<PathBuf, SputnikError>;
+
+    /// returns the globally persistent machine identifier
```
and writes it if not present

jbaxleyiii commented on crates/sputnik/src/session.rs:
```diff
@@ -0,0 +1,130 @@
+#[derive(Debug, Serialize)]
+pub struct Session {
+    /// the command usage where commands are paths and flags are query strings
+    /// i.e. rover profile list --help would become rover/profile/list?help
```
I think this was my original comment and we didn't end up doing this so we should change this comment :)

jbaxleyiii commented on src/bin/rover.rs:
```diff
@@ -1,12 +1,36 @@
+            });
+
+            // kicks off the app on the main thread
+            app.run()?;
```
How would you add custom session information from the invocation of a command (like "this introspection request was for a federated graph") with this setup?

EverlastingBugstopper replied:
gonna mark this one as out of scope/needs further discussion for now. we can revisit later. for now it seems to me like a separate concern and would have a different reporting mechanism entirely as it's quite different from the anonymous usage data we're currently collecting.

jbaxleyiii commented on src/cli.rs:
```diff
@@ -22,3 +29,39 @@ impl Rover {
+    }
+
+    fn is_enabled(&self) -> bool {
+        option_env!("APOLLO_TELEMETRY_DISABLED").is_none()
```
We will need some installer disclosure (as well as docs) about this variable and our usage of telemetry

EverlastingBugstopper replied:
/me nods 

ill add this as a TODO

jbaxleyiii reviewed (commented):
@EverlastingBugstopper I think this is really shaping up nicely. I left some comments about errors that I'm equally happy to trust you and the team on!

My biggest question is around the ability to agument the session report during the invocation of a command like we talked about. I don't see that possible in this current implementation but do think that will be really important for us to have!

EverlastingBugstopper commented on src/cli.rs:
```diff
@@ -22,3 +29,39 @@ impl Rover {
+    fn endpoint(&self) -> Result<Url, SputnikError> {
+        if let Some(url) = option_env!("APOLLO_TELEMETRY_URL") {
+            Ok(url.parse()?)
+        } else if cfg!(debug_assertions) {
```
important to note here that when we're developing on our own machines that metrics will not be reported. want to make sure that's ok behavior

EverlastingBugstopper committed:
feat(houston)!: Convert anyhow to thiserror

EverlastingBugstopper committed:
feat(sputnik): Initial commit

sputnik is a crate that defines an API allowing a CLI tool
to implement their own method of capturing and reporting
anonymous usage data

EverlastingBugstopper committed:
feat(sputnik): Initial commit

sputnik is a crate that defines an API allowing a CLI tool
to implement their own method of capturing and reporting
anonymous usage data

EverlastingBugstopper committed:
feat(rover): Captures anonymous usage data

EverlastingBugstopper committed:
docs(rover): Adds telemetry to CONTRIBUTING.md

EverlastingBugstopper committed:
docs(rover): Adds telemetry documentation

EverlastingBugstopper committed:
tests(rover): tests telemetry reporting

EverlastingBugstopper committed:
test(sputnik): tests r/w machine id

JakeDawkins commented on CONTRIBUTING.md:
```diff
@@ -16,3 +16,32 @@ rover is an open source project developed by [Apollo GraphQL](https://apollograp
+
+When adding a new command or a subcommand, you should make sure that the field in the new command struct is titled `command`. This allows our automated anonymous usage collection to work properly.
+
+Should you add arguments that are _not_ subommands, the arguments you specify will automatically be reported to Apollo's Telemetry backend, unless you specify that an argument should be ignored with `#[serde(skip_serializing)]`. You should be careful to add the `skip_serializing` attribute to any fields that may contain data that may not be anonymous. This can be anything from a profile name to a ...
```
```suggestion
Should you add arguments that are _not_ subcommands, the arguments you specify will automatically be reported to Apollo's Telemetry backend, unless you specify that an argument should be ignored with `#[serde(skip_serializing)]`. You should be careful to add the `skip_serializing` attribute to any fields that may contain data that may not be anonymous. This can be anything from a profile name to a path on the user's computer that could include their name.
```

JakeDawkins reviewed (approved):
This looks really great to me, and I think I actually understand most of it 😂 🎉 

Your documentation work is excellent

StephenBarlow commented on CONTRIBUTING.md:
```diff
@@ -17,3 +17,32 @@ rover is an open source project developed by [Apollo GraphQL](https://apollograp
+
+## Adding new commands/subcommands
+
+When adding a new command or a subcommand, you should make sure that the field in the new command struct is titled `command`. This allows our automated anonymous usage collection to work properly.
```
"you should make sure" -> "make sure"

"This allows our" -> "This enables"

StephenBarlow commented on CONTRIBUTING.md:
```diff
@@ -17,3 +17,32 @@ rover is an open source project developed by [Apollo GraphQL](https://apollograp
+
+When adding a new command or a subcommand, you should make sure that the field in the new command struct is titled `command`. This allows our automated anonymous usage collection to work properly.
+
+Should you add arguments that are _not_ subcommands, the arguments you specify will automatically be reported to Apollo's Telemetry backend, unless you specify that an argument should be ignored with `#[serde(skip_serializing)]`. You should be careful to add the `skip_serializing` attribute to any fields that may contain data that may not be anonymous. This can be anything from a profile name to  ...
```
"Should" -> "If"

"will automatically be reported" -> "are automatically reported"

"You should be careful to" -> "Make sure you"

"that may contain data that may not be anonymous..." -> "that might contain data that isn't anonymous (such as a username or a file path that might include a username)."

StephenBarlow commented on CONTRIBUTING.md:
```diff
@@ -17,3 +17,32 @@ rover is an open source project developed by [Apollo GraphQL](https://apollograp
+
+Should you add arguments that are _not_ subcommands, the arguments you specify will automatically be reported to Apollo's Telemetry backend, unless you specify that an argument should be ignored with `#[serde(skip_serializing)]`. You should be careful to add the `skip_serializing` attribute to any fields that may contain data that may not be anonymous. This can be anything from a profile name to  ...
+
+If you would like to verify that your new command works with the usage data capturing system, you can spin up a development server at `http://localhost:8787/telemetry` (or, if you'd like to use some other endpoint you may specify `APOLLO_TELEMETRY_URL`) and make sure the `command` section of the request body looks right. One example of proper usage collection is when the command `rover config pro ...
```
"If you would like to verify" -> "To verify"

(or, if you'd like...) -> (or specify a different endpoint by setting `APOLLO_TELEMETRY_URL`)

"looks right" -> "looks correct"

New paragraph for "One example..."

"One example..." -> "For example, the usage collection request body for the command `rover config profile show default --sensitive` looks like the following:"

StephenBarlow commented on CONTRIBUTING.md:
````diff
@@ -17,3 +17,32 @@ rover is an open source project developed by [Apollo GraphQL](https://apollograp
+}
+```
+
+As you can see, `default` is left out here in `command.arguments` because it is a user-defined argument and could contain identifiable information, and has been labelled with the `#[serde(skip_serializing)] attribute.
````
"`default` is left out here in" -> "`default` is omitted from"

"could contain" -> "might contain"

Missing a close backtick at the end here I think

StephenBarlow commented on docs/source/privacy/index.md:
```diff
@@ -0,0 +1,20 @@
+
+Rover reports the following data each time you run a command:
+
+- The command that was run _(excluding any identifiable arguments such as file-system paths or profile names)_
```
Don't need to italicize the parenthetical, add a comma after "arguments"

StephenBarlow commented on docs/source/usage/config/environment-variables.md:
```diff
@@ -14,5 +14,7 @@ configuration.
 |--------------------|----------------------------------------------------|-------------------------------------|-------------------------------------------------------------------------------------------------------------------------|
-| APOLLO_CONFIG_HOME | Path to where CLI configuration should be stored.  | Default OS Configuration directory. | More information in [Authentication - Configuration](./authentication#configuration).                  |
-| APOLLO_KEY         | An API Key generated from Apollo Studio.           | none                                | More information in [Authentication - Environment Variables](./authentication#environment-variables). |
+| `APOLLO_CONFIG_HOME` | Path to where CLI configuration should be stored.  | Default OS Configuration directory. | More information in [Authentication - Configuration](./authentication#configuration).                  |
```
"should be" -> "is"

StephenBarlow commented on docs/source/usage/config/environment-variables.md:
```diff
@@ -14,5 +14,7 @@ configuration.
-| APOLLO_KEY         | An API Key generated from Apollo Studio.           | none                                | More information in [Authentication - Environment Variables](./authentication#environment-variables). |
+| `APOLLO_CONFIG_HOME` | Path to where CLI configuration should be stored.  | Default OS Configuration directory. | More information in [Authentication - Configuration](./authentication#configuration).                  |
+| `APOLLO_KEY`         | An API Key generated from Apollo Studio.           | none                                | More information in [Authentication - Environment Variables](./authentication#environment-variables). |
+| `APOLLO_TELEMETRY_URL` | The URL where anonymous usage data should be reported | [https://install.apollographql.com/telemetry](https://install.apollographql.com/telemetry) | More information in [Privacy](../../privacy). |
```
"should be" -> "is"

StephenBarlow commented on docs/source/usage/config/environment-variables.md:
```diff
@@ -14,5 +14,7 @@ configuration.
+| `APOLLO_CONFIG_HOME` | Path to where CLI configuration should be stored.  | Default OS Configuration directory. | More information in [Authentication - Configuration](./authentication#configuration).                  |
+| `APOLLO_KEY`         | An API Key generated from Apollo Studio.           | none                                | More information in [Authentication - Environment Variables](./authentication#environment-variables). |
+| `APOLLO_TELEMETRY_URL` | The URL where anonymous usage data should be reported | [https://install.apollographql.com/telemetry](https://install.apollographql.com/telemetry) | More information in [Privacy](../../privacy). |
+| `APOLLO_TELEMETRY_DISABLED` | 1, if you don't want Rover to collect anonymous usage data | none | |
```
"Set to `1` if you don't want..."

EverlastingBugstopper committed:
Merge pull request #52 from apollographql/avery/test-telemetry

tests telemetry reporting

JakeDawkins reviewed (approved):
🎉 👍
